File: supertux/tileset.py
# Flexlay - A Generic 2D Game Editor
# Copyright (C) 2014 Ingo Ruhnke <grumbel@gmail.com>
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.


from typing import Optional, Any

import os
import logging

from flexlay.tile import Tile
from flexlay.tileset import Tileset
from flexlay.pixel_buffer import PixelBuffer
from flexlay.util.config import Config
from flexlay.util import get_value_from_tree
from supertux.util import load_lisp


class TileGroup:

    def __init__(self, name: str, tiles: list[int]) -> None:
        self.name: str = name
        self.tiles: list[int] = tiles


# Load game tiles from filename into tileset
class SuperTuxTileset(Tileset):

    current: Optional['SuperTuxTileset'] = None

    def __init__(self, *params: Any) -> None:
        super().__init__(*params)
        SuperTuxTileset.current = self
        self.tilegroups: list[TileGroup] = []
        self.filename: str = ""

    def create_ungrouped_tiles_group(self) -> None:
        self.tilegroups.append(TileGroup("Ungrouped Tiles", self.get_ungrouped_tiles()))

    def get_ungrouped_tiles(self) -> list[int]:
        # Searches for tiles which are not yet grouped and creates a group
        # for them
        ungrouped_tiles = []
        for tile in self.get_tiles():
            ungrouped = True
            for group in self.tilegroups:
                if tile not in group.tiles:
                    ungrouped = False
                    break
            if ungrouped:
                ungrouped_tiles.append(tile)
        return ungrouped_tiles

    def load(self, filename: str) -> None:
        logging.info("Loading Tileset: %s" % filename)
        self.filename = filename
        tree: Any = load_lisp(filename, "supertux-tiles")

        tree = tree[1:]

        for i in tree:
            if i[0] == "tiles":
                data = i[1:]
                width = get_value_from_tree(['width', '_'], data, 1)
                # height = get_value_from_tree(['height', '_'], data, 1)
                ids = get_value_from_tree(['ids'], data, [])
                # attributes = get_value_from_tree(['attributes'], data, [])
                image = get_value_from_tree(['image', '_'], data, None)

                if not image:
                    image = get_value_from_tree(['images', '_'], data, None)

                if not image:
                    image = get_value_from_tree(['editor-images', '_'], data, "tiles/auxiliary/notile.png")

                assert Config.current is not None

                x = 0
                y = 0
                for tile_id in ids:
                    pixelbuffer = PixelBuffer.subregion_from_file(os.path.join(Config.current.datadir, "images", image),
                                                                  x * 32, y * 32, 32, 32)
                    self.add_tile(tile_id, Tile(pixelbuffer))
                    x += 1
                    if x == width:
                        x = 0
                        y += 1

            elif i[0] == "tile":
                data = i[1:]
                tile_id = get_value_from_tree(['id', '_'], data, -1)
                image = get_value_from_tree(['editor-images', '_'], data, False)
                hidden = get_value_from_tree(['hidden', '_'], data, False)

                if not image:
                    image = get_value_from_tree(['images', '_'], data, "tiles/auxiliary/notile.png")

                assert Config.current is not None

                if isinstance(image, str):
                    pixelbuffer = PixelBuffer.from_file(
                        os.path.join(Config.current.datadir, "images", image))
                elif isinstance(image, list):
                    if image[0] == "region":
                        pixelbuffer = PixelBuffer.subregion_from_file(
                            os.path.join(Config.current.datadir, "images", image[1]),
                            image[2], image[3], image[4], image[5])

                if not hidden:
                    if tile_id == 0 or not pixelbuffer:
                        self.add_tile(tile_id, None)
                    else:
                        self.add_tile(tile_id, Tile(pixelbuffer))

            elif i[0] == "tilegroup":
                data = i[1:]
                name = get_value_from_tree(['name', '_'], data, "Unnamed")
                tiles = get_value_from_tree(['tiles'], data, [])

                if not self.tilegroups:
                    self.tilegroups = []
                self.tilegroups.append(TileGroup(name, tiles))


# EOF #

File: supertux/tileset.rb
class TileGroup
  attr_accessor :name, :tiles
  
  def initialize(name, tiles)
    @name  = name
    @tiles = tiles
  end 
end

# Load game tiles from filename into tileset
class Tileset
  alias old_initialize initialize

  attr_accessor :tilegroups

  def initialize(*params)
    old_initialize(*params)

    @tilegroups = []
  end

  def load(filename)
    puts "Loading Tileset: #{filename}"
    tree = sexpr_read_from_file(filename)
    if tree == nil then
      puts "Error; Couldn't load: ", filename
      return 
    end

    tree = tree[1..-1]
    counter = 0
    
    tree.each do |i|
      case i[0]
      when "tile"
        data   = i[1..-1]
        id     = get_value_from_tree(['id', '_'], data, -1)
        image  = get_value_from_tree(['editor-images', '_'], data, false)
        hidden = get_value_from_tree(['hidden', '_'], data, false)

        # puts "Loading tile: #{id} => #{image}"

        if not(image)
          image = get_value_from_tree(['images', '_'], data, "notile.png")
        end
        
        if image.is_a?(String) then
          pixelbuffer = make_pixelbuffer($datadir + 'images/tilesets/' + image)
        elsif image.is_a?(Array) then
          if image[0] == "region" then
            pixelbuffer = make_region_pixelbuffer($datadir + 'images/tilesets/' + image[1],
                                                  image[2], image[3], image[4], image[5])
          end
        end
        
        if not hidden then
          if id == 0 then
            add_tile(id, nil)
          else
            add_tile(id, Tile.new(pixelbuffer))
          end
        end

      when "tilegroup"
        data  = i[1..-1]
        name  = get_value_from_tree(['name', '_'], data, "Unnamed")
        tiles = get_value_from_tree(['tiles'], data, [])
        
        if not @tilegroups then
          @tilegroups = []
        end
        @tilegroups.push(TileGroup.new(name, tiles))
      end

      counter += 1
      if counter % 20 == 0 then
        print "Loading tiles: %3.0f%%\r" % [counter.to_f/tree.length.to_f*100.0]
        $stdout.flush()
      end
    end
    puts ""
  end

  def create_ungrouped_tiles_group()
    @tilegroups.push(TileGroup.new("Ungrouped Tiles", get_ungrouped_tiles()))
  end

  def get_ungrouped_tiles()
    # Searches for tiles which are not yet grouped and creates a group
    # for them   
    # Potentially quite slow
    ungrouped_tiles = []
    get_tiles().each {|tile|
      catch :tile_is_grouped do
        tilegroups.each {|group|
          if group.tiles.index(tile) then
            throw :tile_is_grouped
          end
        }
        ungrouped_tiles.push(tile)
      end
    }
    return ungrouped_tiles
  end
end

# EOF #
